File: web/src/video/frame-blitter.ts
/**
 * FrameBlitter — copy a GPUTexture to an ImageBitmap, GPU-resident, placing the
 * source frame into the output canvas per a scale mode + placement transform.
 *
 * Bridges a main-thread VideoPlaybackService output texture across the worker
 * boundary: the engine's render device lives in the engine-worker, so a texture
 * decoded on the main thread can't be sampled there directly. An ImageBitmap is
 * the transferable hand-off (same path the old <video> pump used). This is a
 * straight passthrough — unlike TraceCapture it does NOT checkerboard or force
 * opacity, so the engine receives the frame's exact pixels (straight alpha).
 *
 * The placement is a full-canvas quad whose fragment shader maps each output
 * pixel back to a source UV (honouring the scale-mode footprint, anchor, extra
 * scale, quarter-turn rotation and flips); pixels that fall outside the source
 * are left TRANSPARENT so layers below show through. Orientation matches
 * `createImageBitmap(<video>)`: source top row → bitmap top row.
 */

const BLIT_SHADER = /* wgsl */`
  struct VsOut {
    @builtin(position) pos: vec4f,
    @location(0) uv: vec2f,
  }
  @vertex fn vs(@builtin(vertex_index) i: u32) -> VsOut {
    let x = f32(i32(i) / 2) * 4.0 - 1.0;
    let y = f32(i32(i) % 2) * 4.0 - 1.0;
    var out: VsOut;
    out.pos = vec4f(x, y, 0.0, 1.0);
    // Flip Y: framebuffer origin is top-left, clip-space Y is up → canvas uv
    // (0,0) is the top-left corner.
    out.uv = vec2f((x + 1.0) * 0.5, 1.0 - (y + 1.0) * 0.5);
    return out;
  }
  // rect = (x, y, w, h) destination in normalised canvas coords; rotFlip =
  // (rot 0..3, flipH, flipV, _pad).
  struct Place { rect: vec4f, rotFlip: vec4f };
  @group(0) @binding(0) var src: texture_2d<f32>;
  @group(0) @binding(1) var samp: sampler;
  @group(0) @binding(2) var<uniform> place: Place;
  @fragment fn fs(@location(0) uv: vec2f) -> @location(0) vec4f {
    let r = (uv - place.rect.xy) / place.rect.zw;     // rect-local [0,1]
    let rot = i32(place.rotFlip.x + 0.5);
    var s: vec2f;
    if (rot == 1)      { s = vec2f(r.y, 1.0 - r.x); } // 90° CW
    else if (rot == 2) { s = vec2f(1.0 - r.x, 1.0 - r.y); }
    else if (rot == 3) { s = vec2f(1.0 - r.y, r.x); } // 270° CW
    else               { s = r; }
    if (place.rotFlip.y > 0.5) { s.x = 1.0 - s.x; }
    if (place.rotFlip.z > 0.5) { s.y = 1.0 - s.y; }
    // Sample unconditionally (uniform control flow — no branch around the sample),
    // then mask outside-source pixels to transparent.
    let inside = select(0.0, 1.0, s.x >= 0.0 && s.x <= 1.0 && s.y >= 0.0 && s.y <= 1.0);
    return textureSample(src, samp, clamp(s, vec2f(0.0), vec2f(1.0))) * inside;
  }
`;

/** How the source frame scales into the target canvas. */
export type BlitFit = 'fit' | 'cover' | 'stretch' | 'none';

/** Placement transform layered on the scale-mode fit (see model SourceTransform). */
export interface BlitTransform {
  anchorX: number;
  anchorY: number;
  scale: number;
  rotation: 0 | 90 | 180 | 270;
  flipH: boolean;
  flipV: boolean;
}

const IDENTITY: BlitTransform = { anchorX: 0.5, anchorY: 0.5, scale: 1, rotation: 0, flipH: false, flipV: false };

/** Geometry the placement shader needs: destination rect (normalised canvas) +
 *  quarter-turn index + flips. */
export interface PlaceGeom {
  /** [x, y, w, h] in normalised canvas coords; may extend beyond [0,1] (overflow
   *  → clipped) or start negative (panned crop window). */
  rect: [number, number, number, number];
  /** Quarter-turn index 0..3 (×90° clockwise). */
  rot: number;
  flipH: boolean;
  flipV: boolean;
}

/**
 * Pure placement geometry. `W`/`H` are the blit target (maybe a downscaled
 * preview); `logicalW`/`logicalH` are the size 'none' reasons about — the
 * composition resolution — so "native pixels" is 1:1 against the COMPOSITION
 * then uniformly scaled into the preview. The base footprint per mode is computed
 * on the ROTATED aspect (a 90°/270° frame fits/covers by its turned bounding box),
 * then `scale` zooms about the centre and the anchor positions it; the same
 * `offset = anchor·(canvas − frame)` formula covers both letterbox (frame smaller)
 * and crop (frame larger) regimes.
 */
export function placeGeom(
  sw: number, sh: number, W: number, H: number, mode: BlitFit,
  xf: BlitTransform = IDENTITY, logicalW = W, logicalH = H,
): PlaceGeom {
  const rot = ((((xf.rotation / 90) | 0) % 4) + 4) % 4; // 0..3
  const rotated = rot === 1 || rot === 3;
  const esw = rotated ? sh : sw; // rotated frame's footprint aspect
  const esh = rotated ? sw : sh;
  let dw: number, dh: number;
  if (sw <= 0 || sh <= 0 || mode === 'stretch') {
    dw = W; dh = H;
  } else if (mode === 'cover') {
    const s = Math.max(W / esw, H / esh); dw = esw * s; dh = esh * s;
  } else if (mode === 'none') {
    dw = esw * (W / Math.max(1, logicalW)); dh = esh * (H / Math.max(1, logicalH));
  } else { // fit (contain)
    const s = Math.min(W / esw, H / esh); dw = esw * s; dh = esh * s;
  }
  const scale = Math.max(1e-3, xf.scale);
  dw *= scale; dh *= scale;
  const dx = xf.anchorX * (W - dw); // anchor 0 = left/top edges, 1 = right/bottom
  const dy = xf.anchorY * (H - dh);
  return { rect: [dx / W, dy / H, dw / W, dh / H], rot, flipH: !!xf.flipH, flipV: !!xf.flipV };
}

/** Test hook — the pure placement geometry. */
export const __placeGeomForTest = placeGeom;

export class FrameBlitter {
  private device: GPUDevice;
  private format: GPUTextureFormat = 'rgba8unorm';
  private pipeline: GPURenderPipeline;
  private sampler: GPUSampler;
  private canvas: OffscreenCanvas | null = null;
  private ctx: GPUCanvasContext | null = null;
  private w = 0;
  private h = 0;
  private placeBuf: GPUBuffer;

  constructor(device: GPUDevice) {
    this.device = device;
    this.placeBuf = device.createBuffer({ size: 32, usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST });
    const m = device.createShaderModule({ code: BLIT_SHADER });
    this.pipeline = device.createRenderPipeline({
      layout: 'auto',
      vertex: { module: m, entryPoint: 'vs' },
      fragment: { module: m, entryPoint: 'fs', targets: [{ format: this.format }] },
    });
    this.sampler = device.createSampler({ magFilter: 'linear', minFilter: 'linear' });
  }

  /**
   * Render `srcTexture` (rgba8) to an ImageBitmap of `width × height`, placing the
   * frame per `mode` + `xf`. Areas not covered by the source are left TRANSPARENT
   * so layers below show through.
   */
  toImageBitmap(
    srcTexture: GPUTexture, width: number, height: number, mode: BlitFit = 'stretch',
    xf: BlitTransform = IDENTITY, logicalW = width, logicalH = height,
  ): ImageBitmap {
    if (!this.canvas || this.w !== width || this.h !== height) {
      this.canvas = new OffscreenCanvas(width, height);
      this.ctx = this.canvas.getContext('webgpu') as GPUCanvasContext;
      // premultiplied (not opaque) so transparent areas survive.
      this.ctx.configure({ device: this.device, format: this.format, alphaMode: 'premultiplied' });
      this.w = width;
      this.h = height;
    }
    const g = placeGeom(srcTexture.width, srcTexture.height, width, height, mode, xf, logicalW, logicalH);
    this.device.queue.writeBuffer(this.placeBuf, 0, new Float32Array([
      g.rect[0], g.rect[1], g.rect[2], g.rect[3],
      g.rot, g.flipH ? 1 : 0, g.flipV ? 1 : 0, 0,
    ]));
    const target = this.ctx!.getCurrentTexture();
    const bind = this.device.createBindGroup({
      layout: this.pipeline.getBindGroupLayout(0),
      entries: [
        { binding: 0, resource: srcTexture.createView() },
        { binding: 1, resource: this.sampler },
        { binding: 2, resource: { buffer: this.placeBuf } },
      ],
    });
    const enc = this.device.createCommandEncoder();
    const pass = enc.beginRenderPass({
      colorAttachments: [{
        view: target.createView(),
        loadOp: 'clear',
        storeOp: 'store',
        clearValue: { r: 0, g: 0, b: 0, a: 0 }, // transparent
      }],
    });
    pass.setPipeline(this.pipeline);
    pass.setBindGroup(0, bind);
    pass.draw(3);
    pass.end();
    this.device.queue.submit([enc.finish()]);
    return this.canvas!.transferToImageBitmap();
  }
}

File: native/src/media/frame_blitter.cpp
#include "frame_blitter.h"

#include <algorithm>
#include <cmath>

#include "gpu/gpu_backend.h"

namespace nano_media {
namespace {

/// The MSL twin of frame-blitter.ts's BLIT_SHADER. Web runs it as a
/// full-screen-triangle fragment pass; here it's a compute kernel, which lands
/// on the same pixel centres: the fragment's interpolated uv at pixel p is
/// (p + 0.5)/size, and `gid` addresses the same top-left-origin grid.
constexpr const char* kBlitMSL = R"MSL(
// nano_threadgroup: 8 8 1
#include <metal_stdlib>
using namespace metal;

struct Place { float4 rect; float4 rotFlip; };

kernel void frame_blit(texture2d<float, access::sample> src [[texture(0)]],
                       texture2d<float, access::write>  dst [[texture(1)]],
                       constant Place& place [[buffer(0)]],
                       uint2 gid [[thread_position_in_grid]]) {
  const uint w = dst.get_width();
  const uint h = dst.get_height();
  if (gid.x >= w || gid.y >= h) return;
  const float2 uv = (float2(gid) + 0.5f) / float2(w, h);
  const float2 r = (uv - place.rect.xy) / place.rect.zw;   // rect-local [0,1]
  const int rot = int(place.rotFlip.x + 0.5f);
  float2 s;
  if (rot == 1)      { s = float2(r.y, 1.0f - r.x); }      // 90 CW
  else if (rot == 2) { s = float2(1.0f - r.x, 1.0f - r.y); }
  else if (rot == 3) { s = float2(1.0f - r.y, r.x); }      // 270 CW
  else               { s = r; }
  if (place.rotFlip.y > 0.5f) { s.x = 1.0f - s.x; }
  if (place.rotFlip.z > 0.5f) { s.y = 1.0f - s.y; }
  // Sample unconditionally, then mask outside-source pixels to transparent.
  const float inside = (s.x >= 0.0f && s.x <= 1.0f && s.y >= 0.0f && s.y <= 1.0f) ? 1.0f : 0.0f;
  constexpr sampler samp(coord::normalized, filter::linear, address::clamp_to_edge);
  dst.write(src.sample(samp, clamp(s, float2(0.0f), float2(1.0f))) * inside, gid);
}
)MSL";

/// gpu.h BufferUsage::Uniform.
constexpr int32_t kBufferUsageUniform = 2;

}  // namespace

BlitFit blitFitFromString(const std::string& s) {
  if (s == "cover") return BlitFit::Cover;
  if (s == "stretch") return BlitFit::Stretch;
  if (s == "none") return BlitFit::None;
  return BlitFit::Fit;
}

PlaceGeom placeGeom(int sw, int sh, int W, int H, BlitFit mode, const BlitTransform& xf,
                    int logicalW, int logicalH) {
  PlaceGeom g;
  // TS: ((((xf.rotation / 90) | 0) % 4) + 4) % 4 — truncating divide, then a
  // positive modulo so a negative rotation still lands in 0..3.
  const int turns = (int)(xf.rotation / 90.0);
  g.rot = ((turns % 4) + 4) % 4;
  const bool rotated = g.rot == 1 || g.rot == 3;
  const double esw = rotated ? sh : sw;  // rotated frame's footprint aspect
  const double esh = rotated ? sw : sh;
  double dw = 0, dh = 0;
  if (sw <= 0 || sh <= 0 || mode == BlitFit::Stretch) {
    dw = W; dh = H;
  } else if (mode == BlitFit::Cover) {
    const double s = std::max(W / esw, H / esh);
    dw = esw * s; dh = esh * s;
  } else if (mode == BlitFit::None) {
    dw = esw * (W / (double)std::max(1, logicalW));
    dh = esh * (H / (double)std::max(1, logicalH));
  } else {  // Fit (contain)
    const double s = std::min(W / esw, H / esh);
    dw = esw * s; dh = esh * s;
  }
  const double scale = std::max(1e-3, xf.scale);
  dw *= scale; dh *= scale;
  const double dx = xf.anchorX * (W - dw);  // anchor 0 = left/top edges, 1 = right/bottom
  const double dy = xf.anchorY * (H - dh);
  g.rect[0] = dx / W;
  g.rect[1] = dy / H;
  g.rect[2] = dw / W;
  g.rect[3] = dh / H;
  g.flipH = xf.flipH;
  g.flipV = xf.flipV;
  return g;
}

FrameBlitter::~FrameBlitter() {
  if (!backend_) return;
  if (placeBuf_ >= 0) backend_->release(placeBuf_);
  if (pso_ >= 0) backend_->release(pso_);
  if (shader_ >= 0) backend_->release(shader_);
}

bool FrameBlitter::ensurePipeline(gpu::GPUBackend* backend) {
  if (backend_ && backend_ != backend) return false;
  backend_ = backend;
  if (pso_ < 0) {
    shader_ = backend->createShaderModule(kBlitMSL);
    if (shader_ < 0) return false;
    pso_ = backend->createComputePSO(shader_, "frame_blit");
    if (pso_ < 0) return false;
  }
  if (placeBuf_ < 0) {
    placeBuf_ = backend->createBuffer(32, kBufferUsageUniform);
    if (placeBuf_ < 0) return false;
  }
  return true;
}

bool FrameBlitter::blit(gpu::GPUBackend* backend, int32_t srcTex, int srcW, int srcH,
                        int32_t dstTex, int W, int H, BlitFit mode, const BlitTransform& xf,
                        int logicalW, int logicalH) {
  if (!backend || srcTex < 0 || dstTex < 0 || W <= 0 || H <= 0) return false;
  if (!ensurePipeline(backend)) return false;

  const PlaceGeom g = placeGeom(srcW, srcH, W, H, mode, xf, logicalW, logicalH);
  const float place[8] = {
      (float)g.rect[0], (float)g.rect[1], (float)g.rect[2], (float)g.rect[3],
      (float)g.rot,     g.flipH ? 1.f : 0.f, g.flipV ? 1.f : 0.f, 0.f,
  };
  backend->writeBuffer(placeBuf_, 0, reinterpret_cast<const uint8_t*>(place), sizeof(place));

  const int32_t pass = backend->beginComputePass();
  backend->computeSetPSO(pass, pso_);
  backend->computeSetTexture(pass, srcTex, 0, /*access=*/0);
  backend->computeSetTexture(pass, dstTex, 1, /*access=*/1);
  backend->computeSetBuffer(pass, placeBuf_, 0, /*slot=*/0);
  backend->computeDispatch(pass, (W + 7) / 8, (H + 7) / 8, 1);
  backend->endComputePass(pass);
  return true;
}

}  // namespace nano_media
